sockeye/data_io.py

```diff
@@ -1025,45 +1017,45 @@ def __init__(self,
             check_condition(not add_bos, "Adding a BOS symbol requires a vocabulary")
         self.add_bos = add_bos
         self.limit = limit
-        self._iter = None  # type: Optional[Iterator]
-        self._iterated_once = False
-        self.count = 0
-        self._next = None
 
     def __iter__(self):
-        check_condition(self._next is None, "Can not iterate multiple times simultaneously.")
-        self._iter = read_content(self.path, self.limit)
-        self._next = next(self._iter, None)
-        return self
-
-    def __next__(self):
-        if self._next is None:
-            raise StopIteration
-
-        tokens = self._next
-        if self.vocab is not None:
-            sequence = tokens2ids(tokens, self.vocab)
-        else:
-            sequence = strids2ids(tokens)
-        check_condition(bool(sequence), "Empty sequence in file %s" % self.path)
-
-        if vocab is not None and self.add_bos:
-            sequence.insert(0, self.vocab[C.BOS_SYMBOL])
+        for tokens in read_content(self.path, self.limit):
+            if self.vocab is not None:
+                sequence = tokens2ids(tokens, self.vocab)
+            else:
+                sequence = strids2ids(tokens)
+            if not sequence:
+                yield None
+                continue
+            if vocab is not None and self.add_bos:
+                sequence.insert(0, self.vocab[C.BOS_SYMBOL])
+            yield sequence
 
-        if not self._iterated_once:
-            self.count += 1
 
-        # fetch next element
-        self._next = next(self._iter, None)
-        if self._next is None:
-            self._iter = None
-            if not self._iterated_once:
-                self._iterated_once = True
+def parallel_iter(source_iters: Sequence[Iterable], target_iterable: Iterable):
```

Comment:
we could even make this `Sequence[Iterable[Optional[Any]]]` to denote that we have the expectation that some are none. Anyway, not super important..